### PR TITLE
[pt2] Fix arange to match ATen behavior

### LINKED ISSUE #92676: [inductor] Dtype wrong for arange with floating point step
### 🐛 Describe the bug

```
import torch
from torch.nn import *

class Repro(torch.nn.Module):
    def __init__(self):
        super().__init__()

    def forward(self):
        arange_1 = torch.arange(512, -512, -1.0)
        return arange_1

mod = Repro()
opt_mod = torch.compile(mod)
print(opt_mod().dtype)
print(mod().dtype)
```
```
torch.int64
torch.float32
```

### Versions

N/A

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### PR DESCRIPTION
**Pull request by peterbell10**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #93353

Fixes #92676

`arange` infers the output dtype from the argument types, but in order to reduce
falling back to ATen, inductor preferred to cast whole number float arguments to
int which gave the wrong output dtype. Instead, this decomposes floating point
arange into the prim equivalent for integers.

This also changes the signature of `prims.arange` to

```python
prims.iota(length, *, start, step, **factory_kwargs)
```

which only supports integers arguments. This is done because calculating the
output size from `start, end, step` is surprisingly complex and liable to off by
one errors so should not be duplicated in each backend.

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

### CODE REVIEW

**Review by lezcano — COMMENTED** [review]
Can't this be done at a lowering / ref level?

**Review by ngimel — APPROVED** [review]

File: torch/_refs/__init__.py [hunk 1]
```diff
@@ -4278,20 +4273,79 @@ def arange(
...
+        start=0,
+        step=1,
+        dtype=torch.int64,
         device=device,
-        # pin_memory=pin_memory,
-        requires_grad=requires_grad,
+        requires_grad=False,
     )
+    index = _maybe_convert_to_dtype(index, computation_dtype)
+    result = start + step * index
```
Comment (lezcano) [inline torch/_refs/__init__.py:4343]: Is this alright, or is it better to do it via `lerp` for accuracy?

File: torch/_refs/__init__.py [hunk 2]
```diff
@@ -4278,20 +4273,79 @@ def arange(
...
+        return prims.iota(
+            length,
+            start=xstart,
+            step=xstep,
+            dtype=dtype,
+            device=device,
+            requires_grad=requires_grad,
+        )
+
+    computation_dtype = utils.get_acc_type(dtype, device)
```
Comment (lezcano) [inline torch/_refs/__init__.py:4333]: Is this not the accumulation type used by reductions?

File: torch/_refs/__init__.py (hunk 1 above)
Comment (peterbell10) [reply torch/_refs/__init__.py:4343]: This is the exact expression used in the `arange` kernels:
https://github.com/pytorch/pytorch/blob/3e6978172e86a97dc4c248406d2b2509a837299f/aten/src/ATen/native/cpu/RangeFactoriesKernel.cpp#L33

File: torch/_refs/__init__.py (hunk 2 above)
Comment (peterbell10) [reply torch/_refs/__init__.py:4333]: Yes and t's also used by several other operators including `arange`:
https://github.com/pytorch/pytorch/blob/3e6978172e86a97dc4c248406d2b2509a837299f/aten/src/ATen/native/cpu/RangeFactoriesKernel.cpp#L23

Comment (lezcano) [reply torch/_refs/__init__.py:4333]: Can't this then be done via `utils.reduction_dtypes`?

Comment (peterbell10) [reply torch/_refs/__init__.py:4333]: That doesn't seem to promote to double on cpu, or am I missing something?

Comment (lezcano) [reply torch/_refs/__init__.py:4333]: ag, I see, it does not. I thought `acc_type` had been deprecated in favour of `opmath_type`...

**Review by lezcano — APPROVED** [review]

**Comment by ezyang** [comment]
there is a print in this pr that wasn't removed
